skfda/preprocessing/smoothing/_basis.py

```diff
@@ -0,0 +1,520 @@
+            self (object)
+
+        """
+
+        _check_r_to_r(X)
+
+        self.input_points_ = X.sample_points[0]
+        self.output_points_ = (self.output_points
+                               if self.output_points is not None
+                               else self.input_points_)
+
+        penalty_matrix = self._penalty_matrix()
+
+        # n is the samples
+        # m is the observations
+        # k is the number of elements of the basis
+
+        # Each sample in a column (m x n)
+        data_matrix = X.data_matrix[..., 0].T
+
+        # Each basis in a column
+        basis_values = self.basis.evaluate(self.input_points_).T
+
+        # If no weight matrix is given all the weights are one
+        weight_matrix = (self.weights if self.weights is not None
+                         else np.identity(basis_values.shape[0]))
+
+        # We need to solve the equation
+        # (phi' W phi + lambda * R) C = phi' W Y
+        # where:
+        #  phi is the basis_values
+        #  W is the weight matrix
+        #  lambda the smoothness parameter
+        #  C the coefficient matrix (the unknown)
+        #  Y is the data_matrix
+
+        if(data_matrix.shape[0] > self.basis.nbasis
+           or self.smoothing_parameter > 0):
+
+            # TODO: The penalty could be None (if the matrix is passed)
+            ndegenerated = self.basis._ndegenerated(self._penalty().order)
+
+            method = self._method_function()
+
+            # If the method provides the complete transformation use it
+            method_fit_transform = getattr(method, "fit_transform", None)
+            if method_fit_transform is not None:
+                return method_fit_transform(estimator=self, X=X, y=y)
+
+            # Otherwise the method is used to compute the coefficients
+            coefficients = method(estimator=self,
+                                  basis_values=basis_values,
+                                  weight_matrix=weight_matrix,
+                                  data_matrix=data_matrix,
+                                  penalty_matrix=penalty_matrix,
+                                  ndegenerated=ndegenerated)
+
+        elif data_matrix.shape[0] == self.basis.nbasis:
+            # If the number of basis equals the number of points and no
+            # smoothing is required
+            coefficients = np.linalg.solve(basis_values, data_matrix).T
+
+        else:  # data_matrix.shape[0] < basis.nbasis
+            raise ValueError(f"The number of basis functions "
+                             f"({self.basis.nbasis}) "
+                             f"exceed the number of points to be smoothed "
+                             f"({data_matrix.shape[0]}).")
+
+        fdatabasis = FDataBasis(
+            basis=self.basis, coefficients=coefficients,
+            keepdims=self.keepdims)
+
+        if self.return_basis:
+            return fdatabasis
+        else:
+            return fdatabasis(self.output_points_)
```

Comment:
According to the documentation, an FDataGrid should be returned in this case.
```suggestion
            return fdatabasis.to_grid(eval_points=self.output_points_)
```

Reply:
You are right. I should also add a test for this behaviour.

Reply:
Changed, and added case in docstring in 73ec1d07ea2d6aae469481a3ed30eaa426f4b273.